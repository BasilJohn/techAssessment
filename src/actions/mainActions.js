/**
 * @format
 * @flow strict-local
 */

export const setInputValue = (inputValue: string) => ({
  type: 'SET_INPUT_VALUE',
  payload: inputValue,
});

export const setEmulatorInformation = (isEmulator: string) => ({
  type: 'SET_EMULATOR_INFO',
  payload: isEmulator,
});
