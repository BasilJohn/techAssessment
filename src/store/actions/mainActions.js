/**
 * @format
 * @flow strict-local
 */

const setInputValue = (inputValue: string) => ({
  type: 'SET_INPUT_VALUE',
  payload: inputValue,
});

const setEmulatorInformation = (isEmulator: string) => ({
  type: 'SET_EMULATOR_INFO',
  payload: isEmulator,
});

export { setInputValue, setEmulatorInformation };
