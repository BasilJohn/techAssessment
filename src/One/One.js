/**
 * @format
 * @flow strict-local
 */
import * as React from 'react';
import {View, Text, TouchableOpacity, TextInput} from 'react-native';
import {styles} from './styles';
import {connect} from 'react-redux';
import {setInputValue, setEmulatorInformation} from '../actions/mainActions';
import {bindActionCreators} from 'redux';
import {NativeModules} from 'react-native';
import {NavigationScreenProps} from '@react-navigation/native';

type OneProps = NavigationScreenProps & {
  inputValue: string,
  isEmulator: string,
  setEmulatorInformation: (isEmulator: string) => string,
  setInputValue: (inputValue: String) => string,
};

export class One extends React.Component<OneProps> {
  componentDidMount() {
    NativeModules.EmulatorInfo.isEmulator((err, value) =>
      this.props.setEmulatorInformation(value),
    );
  }

  navigateToScreen = (screenName: string) => {
    this.props.navigation.navigate(screenName);
  };
  changeInputValue = (text: string) => {
    this.props.setInputValue(text) 
  };

  render(): React.Node {
    return (
      <View style={styles.container}>
        <View style={styles.welcomeViewContainer}>
          <Text>Input Data :{this.props.inputValue}</Text>
        </View>
        <View style={styles.welcomeViewContainer}>
          <Text style={styles.noiceText}>
            App is running in Emulator:{this.props.isEmulator}
          </Text>
        </View>
        <View style={styles.buttonContainer}>
          <TouchableOpacity
            onPress={() => this.navigateToScreen('Two')}
            style={styles.button}>
            <Text style={styles.textStyle}>Go to screen two</Text>
          </TouchableOpacity>
        </View>
        <View style={styles.buttonContainer}>
          <TouchableOpacity
            onPress={() => this.navigateToScreen('Three')}
            style={styles.button}>
            <Text style={styles.textStyle}>Go to screen three</Text>
          </TouchableOpacity>
        </View>
        <View style={styles.buttonContainer}>
          <TouchableOpacity
            onPress={() => this.navigateToScreen('utilityValidator')}
            style={styles.button}>
            <Text style={styles.textStyle}>Go to Utility Validator</Text>
          </TouchableOpacity>
        </View>
        <View style={styles.inputContainer}>
          <TextInput
            testID={'txtInputValue'}
            onChangeText={text => this.changeInputValue(text)}
            value={this.props.inputValue}
            placeholder="Type your name here"
            style={styles.input}></TextInput>
        </View>
      </View>
    );
  }
}

const mapStateToProps = state => {
  const {inputValue, isEmulator} = state.main;
  return {inputValue, isEmulator};
};

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      setInputValue,
      setEmulatorInformation,
    },
    dispatch,
  );

export default connect(mapStateToProps, mapDispatchToProps)(One);
