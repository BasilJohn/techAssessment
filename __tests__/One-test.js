/**
 * @format
 */

import 'react-native';
import React from 'react';
import {shallow} from 'enzyme';
import {One} from '../src/One/One';
import mainReducer from '../src/reducers/reducer';
import configureMockStore from 'redux-mock-store';
import {setInputValue} from '../src/actions/mainActions';
import {NativeModules} from 'react-native';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';

const mockStore = configureMockStore([]);
const store = mockStore({
  main: {
    inputValue: '',
    isEmulator: 'false',
  },
});
NativeModules.EmulatorInfo = {
  isEmulator: jest.fn(),
};

const comp = renderer.create(
  <Provider store={store}>
    <One />
  </Provider>,
);
test('Renders Component One Correctly', () => {
  const tree = comp.toJSON();
  expect(tree).toMatchSnapshot();
});

describe('One Actions & Reducers', () => {
  it('Simulate Action', () => {
    let data = 'Basil';

    //Test Action Creators
    store.dispatch(setInputValue(data));
    expect(setInputValue(data)).toEqual({
      payload: data,
      type: 'SET_INPUT_VALUE',
    });

    //Test Reducers
    const afterState = mainReducer((main = {}), setInputValue(data));
    expect(afterState.main.inputValue).toEqual(data);
  });
});

describe('One Functions', () => {
  const props = {
    navigation: {
      navigate: jest.fn(),
    },
    setInputValue: jest.fn(),
  };
  const component = shallow(
    <One
      store={store}
      setInputValue={props.setInputValue}
      navigation={props.navigation}
    />,
  );
  let instance;
  beforeEach(() => {
    instance = component.instance();
  });

  it('Call navigate Two', () => {
    instance.navigateToScreen('Two');
    expect(props.navigation.navigate).toHaveBeenCalled();
    expect(props.navigation.navigate).toHaveBeenCalledWith('Two');
    expect(props.navigation.navigate).toHaveBeenCalledTimes(1);
  });

  it('Call navigate Three', () => {
    instance.navigateToScreen('Three');
    expect(props.navigation.navigate).toHaveBeenCalled();
    expect(props.navigation.navigate).toHaveBeenCalledTimes(2);
    expect(props.navigation.navigate).toHaveBeenCalledWith('Three');
  });

  it('Change input Value', () => {
    instance.changeInputValue('New Text');
    expect(props.setInputValue).toHaveBeenCalled();
    expect(props.setInputValue).toHaveBeenCalledTimes(1);
    expect(props.setInputValue).toHaveBeenCalledWith('New Text');
  });

});
