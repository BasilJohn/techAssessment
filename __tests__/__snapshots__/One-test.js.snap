// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Component One Correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "height": 66.7,
        "justifyContent": "center",
        "paddingRight": 20,
        "width": "100%",
      }
    }
  >
    <Text>
      Input Data :
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "height": 66.7,
        "justifyContent": "center",
        "paddingRight": 20,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "red",
        }
      }
    >
      App is running in Emulator:
    </Text>
  </View>
  <View
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "grey",
        "height": 133.4,
        "marginBottom": 10,
        "width": "90%",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": "100%",
          "justifyContent": "center",
          "opacity": 1,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        Go to screen two
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "grey",
        "height": 133.4,
        "marginBottom": 10,
        "width": "90%",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": "100%",
          "justifyContent": "center",
          "opacity": 1,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        Go to screen three
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "grey",
        "height": 133.4,
        "marginBottom": 10,
        "width": "90%",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": "100%",
          "justifyContent": "center",
          "opacity": 1,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        Go to Utility Validator
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignSelf": "center",
        "height": 133.4,
        "marginBottom": 10,
        "width": "90%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Type your name here"
      rejectResponderTermination={true}
      style={
        Object {
          "borderWidth": 1,
          "height": "100%",
          "width": "100%",
        }
      }
      testID="txtInputValue"
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;
